src/hugr/region.rs

```diff
@@ -0,0 +1,289 @@
+
+use context_iterators::{ContextIterator, IntoContextIterator, MapWithCtx};
+use itertools::{Itertools, MapInto};
+use portgraph::{Hierarchy, LinkView, MultiPortGraph, PortView, UnmanagedDenseMap};
+
+use crate::{ops::OpType, Direction, Hugr, Node, Port};
+
+use super::HugrView;
+
+type FlatRegionGraph<'g> = portgraph::view::FlatRegion<'g, MultiPortGraph>;
+
+/// Single region view of a HUGR. Ignores any nodes that are not direct children
+/// of the root.
+///
+/// For a view that includes all the descendants of the root, see [`RegionView`].
+#[derive(Clone, Debug)]
+pub struct FlatRegionView<'g> {
+    /// The chosen root node.
+    root: Node,
+
+    /// The graph encoding the adjacency structure of the HUGR.
+    graph: FlatRegionGraph<'g>,
+
+    /// The node hierarchy.
+    hierarchy: &'g Hierarchy,
+
+    /// Operation types for each node.
+    op_types: &'g UnmanagedDenseMap<portgraph::NodeIndex, OpType>,
+}
+
+impl<'g> FlatRegionView<'g> {
+    /// Create a new region view of a HUGR containing only the direct children
+    /// of the root.
+    pub fn new(hugr: &'g Hugr, root: Node) -> Self {
+        let Hugr {
+            graph,
+            hierarchy,
+            op_types,
+            ..
+        } = hugr;
+        Self {
+            root,
+            graph: FlatRegionGraph::new_flat_region(graph, hierarchy, root.index),
+            hierarchy,
+            op_types,
+        }
+    }
+}
+
+impl<'g> HugrView for FlatRegionView<'g> {
+    type Nodes<'a> = MapInto<portgraph::hierarchy::Children<'a>, Node>
+    where
+        Self: 'a;
+
+    type NodePorts<'a> = MapInto<<FlatRegionGraph<'g> as PortView>::NodePortOffsets<'a>, Port>
+    where
+        Self: 'a;
+
+    type Children<'a> = MapInto<portgraph::hierarchy::Children<'a>, Node>
+    where
+        Self: 'a;
+
+    type Neighbours<'a> = MapInto<<FlatRegionGraph<'g> as LinkView>::Neighbours<'a>, Node>
+    where
+        Self: 'a;
+
+    type PortLinks<'a> = MapWithCtx<
+        <FlatRegionGraph<'g> as LinkView>::PortLinks<'a>,
+        &'a Self,
+        (Node, Port),
+    > where
+        Self: 'a;
+
+    fn root(&self) -> Node {
+        self.root
+    }
+
+    fn get_parent(&self, node: Node) -> Option<Node> {
+        self.hierarchy
+            .parent(node.index)
+            .map(Into::into)
+            .filter(|&n| n == self.root)
+    }
+
+    fn get_optype(&self, node: Node) -> &OpType {
+        self.op_types.get(node.index)
+    }
+
+    fn node_count(&self) -> usize {
+        self.hierarchy.child_count(self.root.index) + 1
+    }
+
+    fn edge_count(&self) -> usize {
+        // Faster implementation than filtering all the nodes in the internal graph.
+        self.nodes()
+            .map(|n| self.output_neighbours(n).count())
+            .sum()
+    }
+
+    fn nodes(&self) -> Self::Nodes<'_> {
+        // Faster implementation than filtering all the nodes in the internal graph.
+        self.hierarchy.children(self.root.index).map_into()
```

Comment:
this doesn't include the root? But `node_count` does?

Reply:
Good catch, I updated it and added some simple tests.